src/OmniSharp.Plugins/PluginAssemblies.cs

```diff
@@ -1,16 +1,89 @@
 using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+#if !NET451
+using System.Runtime.Loader;
+#endif
+using Microsoft.Extensions.DependencyModel;
+using Microsoft.Extensions.PlatformAbstractions;
 
 namespace OmniSharp.Plugins
 {
     public class PluginAssemblies
     {
-        private readonly IEnumerable<string> _assemblyNames;
+        private readonly IEnumerable<string> _paths;
+        private readonly IEnumerable<string> _plugins;
+        private ImmutableArray<Assembly> _assemblies;
 
-        public PluginAssemblies(IEnumerable<string> assemblyNames)
+        public PluginAssemblies(IEnumerable<string> paths)
         {
-            _assemblyNames = assemblyNames;
+            _paths = paths;
+            _plugins = _paths
+                .SelectMany(x => Directory.GetFiles(x, "*.dll", SearchOption.TopDirectoryOnly))
+                .Where(x => !x.Contains(".Native."))
+                .ToArray();
         }
 
-        public IEnumerable<string> AssemblyNames { get { return _assemblyNames; } }
+        public IEnumerable<string> Plugins => this._plugins;
+
+        public ImmutableArray<Assembly> Assemblies
+        {
+            get
+            {
+                if (this._assemblies == null)
+                {
+                    this._assemblies = ImmutableArray.CreateRange(
+                        this._plugins
+#if NET451
+                            .Select(AssemblyName.GetAssemblyName)
+                            .Select(Assembly.Load)
+#else
+                            .Select(path =>
+                            {
+                                var loader = new AssemblyLoader(Path.GetDirectoryName(path));
+                                return loader.LoadFromAssemblyPath(path);
+                            })
+#endif
+                            .ToArray()
```

Comment:
`ImmutableArray<T>` is not lazy. It necessarily iterates the `IEnumerable<T>` passed in and creates an array out of it. It won't simply use an array that's passed in because modifications to that array would be observable. So, the `ToArray()` here is an unnecessary allocation.